### ISSUE TITLE
doctest.h -- error: call of overloaded ‘log10(unsigned int)’ is ambiguous

### ISSUE THREAD

**Issue by nbriggs** [doctest/doctest#722]
## Description
Using doctest 2.4.9, as packaged with ccache, compiling with gcc on solaris (see versions below) results in:

```
In file included from /export/home/briggs/ccache-4.7.1/unittest/main.cpp:27:
/export/home/briggs/ccache-4.7.1/src/third_party/doctest.h: In member function ‘virtual void doctest::{anonymous}::ConsoleReporter::test_run_end(const doctest::TestRunStats&)’:
/export/home/briggs/ccache-4.7.1/src/third_party/doctest.h:6191:48: error: call of overloaded ‘log10(unsigned int)’ is ambiguous
 6191 |             auto totwidth = int(std::ceil(log10((std::max(p.numTestCasesPassingFilters, static_cast<unsigned>(p.numAsserts))) + 1)));
      |                                           ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/gcc/11/lib/gcc/sparcv9-sun-solaris2.11/11.2.0/include-fixed/math.h:24,
                 from /usr/gcc/11/include/c++/11.2.0/bits/std_abs.h:40,
                 from /usr/gcc/11/include/c++/11.2.0/cstdlib:77,
                 from /usr/gcc/11/include/c++/11.2.0/ext/string_conversions.h:41,
                 from /usr/gcc/11/include/c++/11.2.0/bits/basic_string.h:6607,
                 from /usr/gcc/11/include/c++/11.2.0/string:55,
                 from /export/home/briggs/ccache-4.7.1/src/Stat.hpp:29,
                 from /export/home/briggs/ccache-4.7.1/unittest/../src/Util.hpp:21,
                 from /export/home/briggs/ccache-4.7.1/unittest/main.cpp:19:
/usr/gcc/11/lib/gcc/sparcv9-sun-solaris2.11/11.2.0/include-fixed/iso/math_iso.h:209:28: note: candidate: ‘long double std::log10(long double)’
  209 |         inline long double log10(long double __X) { return __log10l(__X); }
      |                            ^~~~~
/usr/gcc/11/lib/gcc/sparcv9-sun-solaris2.11/11.2.0/include-fixed/iso/math_iso.h:170:22: note: candidate: ‘float std::log10(float)’
  170 |         inline float log10(float __X) { return __log10f(__X); }
      |                      ^~~~~
/usr/gcc/11/lib/gcc/sparcv9-sun-solaris2.11/11.2.0/include-fixed/iso/math_iso.h:70:15: note: candidate: ‘double std::log10(double)’
   70 | extern double log10 __P((double));
      |               ^~~~~
In file included from /export/home/briggs/ccache-4.7.1/unittest/main.cpp:27:
/export/home/briggs/ccache-4.7.1/src/third_party/doctest.h:6192:49: error: call of overloaded ‘log10(unsigned int)’ is ambiguous
 6192 |             auto passwidth = int(std::ceil(log10((std::max(p.numTestCasesPassingFilters - p.numTestCasesFailed, static_cast<unsigned>(p.numAsserts - p.numAssertsFailed))) + 1)));
      |                                            ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/gcc/11/lib/gcc/sparcv9-sun-solaris2.11/11.2.0/include-fixed/math.h:24,
                 from /usr/gcc/11/include/c++/11.2.0/bits/std_abs.h:40,
                 from /usr/gcc/11/include/c++/11.2.0/cstdlib:77,
                 from /usr/gcc/11/include/c++/11.2.0/ext/string_conversions.h:41,
                 from /usr/gcc/11/include/c++/11.2.0/bits/basic_string.h:6607,
                 from /usr/gcc/11/include/c++/11.2.0/string:55,
                 from /export/home/briggs/ccache-4.7.1/src/Stat.hpp:29,
                 from /export/home/briggs/ccache-4.7.1/unittest/../src/Util.hpp:21,
                 from /export/home/briggs/ccache-4.7.1/unittest/main.cpp:19:
/usr/gcc/11/lib/gcc/sparcv9-sun-solaris2.11/11.2.0/include-fixed/iso/math_iso.h:209:28: note: candidate: ‘long double std::log10(long double)’
  209 |         inline long double log10(long double __X) { return __log10l(__X); }
      |                            ^~~~~
/usr/gcc/11/lib/gcc/sparcv9-sun-solaris2.11/11.2.0/include-fixed/iso/math_iso.h:170:22: note: candidate: ‘float std::log10(float)’
  170 |         inline float log10(float __X) { return __log10f(__X); }
      |                      ^~~~~
/usr/gcc/11/lib/gcc/sparcv9-sun-solaris2.11/11.2.0/include-fixed/iso/math_iso.h:70:15: note: candidate: ‘double std::log10(double)’
   70 | extern double log10 __P((double));
      |               ^~~~~
In file included from /export/home/briggs/ccache-4.7.1/unittest/main.cpp:27:
/export/home/briggs/ccache-4.7.1/src/third_party/doctest.h:6193:49: error: call of overloaded ‘log10(unsigned int)’ is ambiguous
 6193 |             auto failwidth = int(std::ceil(log10((std::max(p.numTestCasesFailed, static_cast<unsigned>(p.numAssertsFailed))) + 1)));
      |                                            ~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /usr/gcc/11/lib/gcc/sparcv9-sun-solaris2.11/11.2.0/include-fixed/math.h:24,
                 from /usr/gcc/11/include/c++/11.2.0/bits/std_abs.h:40,
                 from /usr/gcc/11/include/c++/11.2.0/cstdlib:77,
                 from /usr/gcc/11/include/c++/11.2.0/ext/string_conversions.h:41,
                 from /usr/gcc/11/include/c++/11.2.0/bits/basic_string.h:6607,
                 from /usr/gcc/11/include/c++/11.2.0/string:55,
                 from /export/home/briggs/ccache-4.7.1/src/Stat.hpp:29,
                 from /export/home/briggs/ccache-4.7.1/unittest/../src/Util.hpp:21,
                 from /export/home/briggs/ccache-4.7.1/unittest/main.cpp:19:
/usr/gcc/11/lib/gcc/sparcv9-sun-solaris2.11/11.2.0/include-fixed/iso/math_iso.h:209:28: note: candidate: ‘long double std::log10(long double)’
  209 |         inline long double log10(long double __X) { return __log10l(__X); }
      |                            ^~~~~
/usr/gcc/11/lib/gcc/sparcv9-sun-solaris2.11/11.2.0/include-fixed/iso/math_iso.h:170:22: note: candidate: ‘float std::log10(float)’
  170 |         inline float log10(float __X) { return __log10f(__X); }
      |                      ^~~~~
/usr/gcc/11/lib/gcc/sparcv9-sun-solaris2.11/11.2.0/include-fixed/iso/math_iso.h:70:15: note: candidate: ‘double std::log10(double)’
   70 | extern double log10 __P((double));
      |               ^~~~~
```

### Steps to reproduce

For ccache 4.7.1, doing
```
mkdir build && cd build && cmake --build .
```

### Extra information

* doctest version: 2.4.9
* Operating System: Solaris 11.4
* Compiler+version: gcc 11.2.0

There is indeed no log10() with an integral value argument.  I believe that was introduced in C++11, and I suspect that the Solaris `iso/math_iso.h` predates that.  While Oracle ought to fix that, it seems that a cast to float would simply avoid this dependency.

**Comment by nbriggs** [comment]
(transferred from ccache/ccache#1194)

**Comment by navinp0304** [comment]
This is because in /usr/include/c++/12/cmath you have this line
```
  template<typename _Tp>
    inline _GLIBCXX_CONSTEXPR
    typename __gnu_cxx::__enable_if<__is_integer<_Tp>::__value,
                                    double>::__type
    log10(_Tp __x)
    { return __builtin_log10(__x); }
```
Probably in solaris you don't have that line. I don't have a solaris system to check. But commenting out the above lines gives a error message similar to yours.

An easy fix in doctest would be to change the first arg of log10 to double by static_cast<double> since log10 expect double and the above snipped intrinsically does that.

**Comment by onqtam** [comment]
seems like they also suggest the same: https://github.com/ccache/ccache/issues/1194#issuecomment-1292647079

**Comment by navinp0304** [comment]
> seems like they also suggest the same: [ccache/ccache#1194 (comment)](https://github.com/ccache/ccache/issues/1194#issuecomment-1292647079)

I think we should fix it doctest as well, so that other platforms,static analysis tools don't complain.Already the second argument of max is a static_cast<unsigned>

**Comment by navinp0304** [comment]
Fixed with PR #756
